pkg/netstore/netstore.go

```diff
@@ -57,31 +66,99 @@ func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Addres
 				go s.recoveryCallback(addr, targets)
 				return nil, ErrRecoveryAttempt
 			}
-			stamp, err := ch.Stamp().MarshalBinary()
-			if err != nil {
-				return nil, err
-			}
-
-			putMode := storage.ModePutRequest
-			if mode == storage.ModeGetRequestPin {
-				putMode = storage.ModePutRequestPin
-			}
-
-			cch, err := s.validStamp(ch, stamp)
-			if err != nil {
-				// if a chunk with an invalid postage stamp was received
-				// we force it into the cache.
-				putMode = storage.ModePutRequestCache
-				cch = ch
-			}
-
-			_, err = s.Storer.Put(ctx, putMode, cch)
-			if err != nil {
-				return nil, fmt.Errorf("netstore retrieve put: %w", err)
+			select {
+			case <-ctx.Done():
+			case s.opChan <- putOp{ch: ch, mode: mode}:
 			}
 			return ch, nil
 		}
 		return nil, fmt.Errorf("netstore get: %w", err)
 	}
 	return ch, nil
 }
+
+// The underlying store is not the netstore's responsibility to close
+func (s *store) Close() error {
+	// for some reason if the backgroundPutter was not started, this should be a noop
+	if s.sCancel == nil {
+		return nil
+	}
+
+	s.sCancel()
+
+	stopped := make(chan struct{})
+	go func() {
+		defer close(stopped)
+		s.wg.Wait()
+	}()
+
+	select {
+	case <-stopped:
+		return nil
+	case <-time.After(time.Second * 5):
+		return errors.New("netstore: waited 5 seconds to close active goroutines")
+	}
+}
+
+type putOp struct {
+	ch   swarm.Chunk
+	mode storage.ModeGet
+}
+
+func (s *store) backgroundPutter() {
```

Comment:
I think that if you went as far as introducing an extra worker that does the `Put`s in the background, wouldn't it make more sense to just do write coalescing? If you're spawning a goroutine for every `Put`, you could just avoid the whole complexity of the background worker in favor of a simple async write within the body of the `Get`(potentially with a semaphore embedded into it), which simply ignores the error:
```go
go func() { storer.Put() } ()
```
Write coalescing in this context would be just a period flushing of chunks collected to be put over a period of time.